src/com/xilinx/rapidwright/interchange/DeviceResourcesWriter.java

```diff
@@ -772,10 +809,55 @@ public static Map<TileTypeEnum, Integer> writeAllTileTypesToBuilder(Design desig
+                        boolean includeBase = true;
+                        boolean includeDiscontinuity = false;
+                        float delayPs = RouterHelper.computeNodeDelay(delayEstimator, endNode, includeBase, includeDiscontinuity);
+                        delayPs += DelayEstimatorBase.getExtraDelay(endNode, DelayEstimatorBase.isLong(startNode));
+                        if (pip.isRouteThru()) {
+                            if (Utils.isCLB(pip.getTile().getTileTypeEnum())) {
+                                // Site port to cell
+                                com.xilinx.rapidwright.device.SitePin fromSitePin = pip.getStartWire().getSitePin();
+                                String fromSitePinName = fromSitePin.getPinName();
+                                assert(fromSitePinName.matches("[A-H][1-6]"));
+                                char lutLetter = fromSitePinName.charAt(0);
+                                char pinNumber = fromSitePinName.charAt(1);
+                                delayPs += intrasiteAndLogicDelayModel.getIntraSiteDelay(SiteTypeEnum.SLICEL, fromSitePinName, lutLetter + "6LUT/A" + pinNumber);
+                                // Through cell
+                                short belIdx = intrasiteAndLogicDelayModel.getBELIndex(lutLetter + "6LUT");
+                                delayPs += intrasiteAndLogicDelayModel.getLogicDelay(belIdx, "A" + pinNumber, "O6");
+                                // Cell to site port
+                                com.xilinx.rapidwright.device.SitePin toSitePin = pip.getEndWire().getSitePin();
+                                String toSitePinName = toSitePin.getPinName();
+                                assert(toSitePinName.matches("[A-H](_O|MUX)"));
+                                delayPs += intrasiteAndLogicDelayModel.getIntraSiteDelay(SiteTypeEnum.SLICEL, lutLetter + "6LUT/O6", toSitePinName);
+                            }
+                        }
+                        if (delayPs != 0) {
+                            int index = slowMaxDelayIndices.computeIfAbsent(delayPs, (v) -> slowMaxDelayIndices.size());
+                            pipBuilder.setTiming(index);
+                        }
+                    }
+                }
             }
+
             i++;
         }
 
+        StructList.Builder<DeviceResources.Device.PIPTiming.Builder> pipTimingsBuilder = devBuilder.initPipTimings(slowMaxDelayIndices.size());
+        for (Map.Entry<Float,Integer> e : slowMaxDelayIndices.entrySet()) {
+            float slowMaxDelayPs = e.getKey();
+            int index = e.getValue();
+            DeviceResources.Device.PIPTiming.Builder timingBuilder = pipTimingsBuilder.get(index);
+            DeviceResources.Device.CornerModel.Builder delayBuilder = timingBuilder.initInternalDelay();
+            DeviceResources.Device.CornerModelValues.Builder slowBuilder = delayBuilder.initSlow().initSlow();
+            slowBuilder.initMax().setMax(slowMaxDelayPs * 1e-12f);
```

Comment:
Perhaps pull out `1e-12f` and other occurrences into a separate static variable with a note.